### PR TITLE
Encrypt master key with a password before storing

### LINKED ISSUE #18: Encrypt master key with a password before storing
**Note:** This issue is not about supporting bip39 passphrase, but about wallet encryption. The seed stored in the wallet-db should be encrypted with a password.

- [ ] Install "@types/bip38" as dev-dependency
- [ ] Password should be optional in `init()`. If no password is passed, use the default password `12345678`.
- [ ] Use "bip38" lib to do encryption/decryption.
- [ ] Implement `async setPassword(newPassword, oldPassword): void` in Wallet class to add or update password

### PR DESCRIPTION
**Pull request by emjshrx**
fixes #18

### CODE REVIEW

File: src/wallet/wallet.ts
Comment (theanmolsharma) [inline src/wallet/wallet.ts:1]: A couple of things we should do:

- To prevent bloating up the `init()` let's break it down into smaller methods and call it in `init()`
    1. create a new method: `private async initializeMasterKey(mnemonic?: string, password = '12345678')`
    2. move everything after `await this.db.open();` to that method
- We should also encrypt the `chaincode` before storing it because `chaincode` is also sensitive data.

File: src/wallet/db/db.interface.ts
Comment (theanmolsharma) [inline src/wallet/db/db.interface.ts:12]: I think both arguments should be of the same type.

File: src/wallet/wallet.ts
Comment (theanmolsharma) [inline src/wallet/wallet.ts:281]: Let's move this above `deriveAddress()`

Comment (theanmolsharma) [inline src/wallet/wallet.ts:281]: nit: always leave a blank like between methods

**Review by theanmolsharma — CHANGES_REQUESTED** [review]
Before we merge, I want to see some tests for the functionality you are adding.
Don't merge the PR after addressing the comments, wait for at least 1 approval.

File: src/wallet/wallet.ts
Comment (emjshrx) [reply src/wallet/wallet.ts:1]: instead of adding a new method that will only be used once, I reused `setPassword` in init as well as the code is the same.

File: src/wallet/db/level/db.ts
Comment (emjshrx) [inline src/wallet/db/level/db.ts:41]: Removed return type, as its redundant since already declared in `src/wallet/db/db.interface.ts`

File: src/wallet/wallet.ts
Comment (theanmolsharma) [reply src/wallet/wallet.ts:1]: i think it's fine for now, thanks.

File: src/wallet/wallet.ts
Comment (theanmolsharma) [inline src/wallet/wallet.ts:72]: ley's rename to `setPassword`

File: test/wallet.spec.ts
Comment (theanmolsharma) [inline test/wallet.spec.ts:172]: Don't think we need a new describe block just to test one feature. Just use the setup I have made earlier in the file. This patch should do it.

```
diff --git a/test/wallet.spec.ts b/test/wallet.spec.ts
index 3b172f7..f1a4081 100644
--- a/test/wallet.spec.ts
+++ b/test/wallet.spec.ts
@@ -38,6 +38,13 @@ describe('Wallet', () => {
         });
     });
 
+    it('should set a new password, close and reopen the wallet with the same password', async () => {
+        const password = 'notSoSecretPassword';
+        await wallet.writeMasterKey(password);
+        await wallet.close();
+        await wallet.init({ password });
+    });
+
     it('should derive first receive address', async () => {
         address = await wallet.deriveReceiveAddress();
         expect(address).toBe('bcrt1qcr8te4kr609gcawutmrza0j4xv80jy8zeqchgx');
```

**Review by theanmolsharma — CHANGES_REQUESTED** [review]
Address comments, then we can merge. Nice work!!

File: test/wallet.spec.ts
Comment (emjshrx) [reply test/wallet.spec.ts]: This will not work as the `it` blocks are run concurrently and the tests executed when the wallet is closed fail. This is a flaky test and should be avoided imo.
Screenshot 2024-02-10 at 1 29 50 PM

Hence I separated this out to a separate wallet instance. Maybe we can have two wallet instances in this `describe` block, but again two `describe` blocks seemed cleaner.

Comment (theanmolsharma) [reply test/wallet.spec.ts]: The `it` blocks in a `describe` run sequentially, always. It's the `describe` that runs concurrently.

Comment (emjshrx) [reply test/wallet.spec.ts]: Yes my initial RCA was wrong, this is an issue with the `wallet.close()` method. Thanks for helping with the debugging @theanmolsharma

**Review by theanmolsharma — APPROVED** [review]

**Pull request merged**
